// vue-tailwind/dist/l10n/index.js
import { Arabic } from './ar.js';
import { Austria } from './at.js';
import { Azerbaijan } from './az.js';
import { Belarusian } from './be.js';
import { Bulgarian } from './bg.js';
import { Bangla } from './bn.js';
import { Bosnian } from './bs.js';
import { Catalan } from './cat.js';
import { Czech } from './cs.js';
import { Welsh } from './cy.js';
import { Danish } from './da.js';
import { German } from './de.js';
import { English } from './default.js';
import { Esperanto } from './eo.js';
import { Spanish } from './es.js';
import { Estonian } from './et.js';
import { Persian } from './fa.js';
import { Finnish } from './fi.js';
import { Faroese } from './fo.js';
import { French } from './fr.js';
import { Greek } from './gr.js';
import { Hebrew } from './he.js';
import { Hindi } from './hi.js';
import { Croatian } from './hr.js';
import { Hungarian } from './hu.js';
import { Indonesian } from './id.js';
import { Icelandic } from './is.js';
import { Italian } from './it.js';
import { Japanese } from './ja.js';
import { Georgian } from './ka.js';
import { Korean } from './ko.js';
import { Khmer } from './km.js';
import { Kazakh } from './kz.js';
import { Lithuanian } from './lt.js';
import { Latvian } from './lv.js';
import { Macedonian } from './mk.js';
import { Mongolian } from './mn.js';
import { Malaysian } from './ms.js';
import { Burmese } from './my.js';
import { Dutch } from './nl.js';
import { Norwegian } from './no.js';
import { Punjabi } from './pa.js';
import { Polish } from './pl.js';
import { Portuguese } from './pt.js';
import { Romanian } from './ro.js';
import { Russian } from './ru.js';
import { Sinhala } from './si.js';
import { Slovak } from './sk.js';
import { Slovenian } from './sl.js';
import { Albanian } from './sq.js';
import { Serbian } from './sr.js';
import { Swedish } from './sv.js';
import { Thai } from './th.js';
import { Turkish } from './tr.js';
import { Ukrainian } from './uk.js';
import { Uzbek } from './uz.js';
import { UzbekLatin } from './uz_latn.js';
import { Vietnamese } from './vn.js';
import { Mandarin } from './zh.js';
import { MandarinTraditional } from './zh-tw.js';

const l10n = {
    ar: Arabic,
    at: Austria,
    az: Azerbaijan,
    be: Belarusian,
    bg: Bulgarian,
    bn: Bangla,
    bs: Bosnian,
    ca: Catalan,
    cat: Catalan,
    cs: Czech,
    cy: Welsh,
    da: Danish,
    de: German,
    default: Object.assign({}, English),
    en: English,
    eo: Esperanto,
    es: Spanish,
    et: Estonian,
    fa: Persian,
    fi: Finnish,
    fo: Faroese,
    fr: French,
    gr: Greek,
    he: Hebrew,
    hi: Hindi,
    hr: Croatian,
    hu: Hungarian,
    id: Indonesian,
    is: Icelandic,
    it: Italian,
    ja: Japanese,
    ka: Georgian,
    ko: Korean,
    km: Khmer,
    kz: Kazakh,
    lt: Lithuanian,
    lv: Latvian,
    mk: Macedonian,
    mn: Mongolian,
    ms: Malaysian,
    my: Burmese,
    nl: Dutch,
    no: Norwegian,
    pa: Punjabi,
    pl: Polish,
    pt: Portuguese,
    ro: Romanian,
    ru: Russian,
    si: Sinhala,
    sk: Slovak,
    sl: Slovenian,
    sq: Albanian,
    sr: Serbian,
    sv: Swedish,
    th: Thai,
    tr: Turkish,
    uk: Ukrainian,
    vn: Vietnamese,
    zh: Mandarin,
    zh_tw: MandarinTraditional,
    uz: Uzbek,
    uz_latn: UzbekLatin,
};

export default l10n;
